cmd/containerboot/ingressservices.go

```diff
@@ -0,0 +1,329 @@
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+	if len(j) == 0 || string(j) == "" {
+		return nil, nil
+	}
+	cfg := &ingressservices.Configs{}
+	if err := json.Unmarshal(j, &cfg); err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+// getStatus gets the recorded status of the configured firewall. The status is
+// stored in the proxy's state Secret.  Note that the recorded status might not
+// be the current status of the firewall if it belongs to a previous Pod- we
+// take that into account further down the line when determining if the desired
+// rules are actually present.
+func (ep *ingressProxy) getStatus(ctx context.Context) (*ingressservices.Status, error) {
+	secret, err := ep.kc.GetSecret(ctx, ep.stateSecret)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving state secret: %w", err)
+	}
+	status := &ingressservices.Status{}
+	raw, ok := secret.Data[ingressservices.IngressConfigKey]
+	if !ok {
+		return nil, nil
+	}
+	if err := json.Unmarshal([]byte(raw), status); err != nil {
+		return nil, fmt.Errorf("error unmarshalling previous config: %w", err)
+	}
+	return status, nil
+}
+
+// syncIngressConfigs takes the desired firewall configuration and the recorded
+// status and ensures that any missing rules are added and no longer needed
+// rules are deleted.
+func (ep *ingressProxy) syncIngressConfigs(cfgs *ingressservices.Configs, status *ingressservices.Status) error {
+	rulesToAdd := ep.getRulesToAdd(cfgs, status)
+	rulesToDelete := ep.getRulesToDelete(cfgs, status)
+
+	if err := ensureIngressRulesDeleted(rulesToDelete, ep.nfr); err != nil {
+		return fmt.Errorf("error deleting ingress rules: %w", err)
+	}
+	if err := ensureIngressRulesAdded(rulesToAdd, ep.nfr); err != nil {
+		return fmt.Errorf("error adding ingress rules: %w", err)
+	}
+	return nil
+}
+
+// recordStatus writes the configured firewall status to the proxy's state
+// Secret. This allows the Kubernetes Operator to determine whether this proxy
+// Pod has setup firewall rules to route traffic for an ingress service.
+func (ep *ingressProxy) recordStatus(ctx context.Context, newCfg *ingressservices.Configs) error {
+	status := &ingressservices.Status{}
+	if newCfg != nil {
+		status.Configs = *newCfg
+	}
+	// Pod IPs are used to determine if recorded status applies to THIS proxy Pod.
+	status.PodIPv4 = ep.podIPv4
+	status.PodIPv6 = ep.podIPv6
+	secret, err := ep.kc.GetSecret(ctx, ep.stateSecret)
+	if err != nil {
+		return fmt.Errorf("error retrieving state Secret: %w", err)
+	}
+	bs, err := json.Marshal(status)
+	if err != nil {
+		return fmt.Errorf("error marshalling status: %w", err)
+	}
+	secret.Data[ingressservices.IngressConfigKey] = bs
+	patch := kubeclient.JSONPatch{
+		Op:    "replace",
+		Path:  fmt.Sprintf("/data/%s", ingressservices.IngressConfigKey),
+		Value: bs,
+	}
+	if err := ep.kc.JSONPatchResource(ctx, ep.stateSecret, kubeclient.TypeSecrets, []kubeclient.JSONPatch{patch}); err != nil {
```

Comment:
Would it be possible/sensible to use update instead? This is running in a goroutine and is not the only code that updates the state Secret so it feels a bit vulnerable to subtle overwrite bugs.

Reply:
I think we can be quite certain that the patch operation itself will work as expected (will only rewrite the value of the key set in the patch) - if that was broken it would be Kubernetes-side/in the jsonpatch library somehow. And the key that we set is the ingress config key that is only set by the ingress config syncer here that runs to finish synchronously before waiting for other events - so I think we should be concurrency safe

Reply:
Oh yeah, good point about the specific key, I missed that detail 👍